### ISSUE TITLE
Unexpected error when 2 provided args of same type are present

### ISSUE THREAD

**Issue by btilbrook-nextfaze** [square/AssistedInject#46]
Code similar to this:

```kotlin
class MyClass @AssistedInject constructor(
    foobar: Foobar,
    @Assisted a: String,
    @Assisted b: String
) {
    @AssistedInject.Factory
    interface Factory {
        fun create(a: String, b: String): MyClass
    }
}
```

Results in:

```
Duplicate @Assisted parameters declared. Forget a qualifier annotation?
```

Since `@Assisted` args are supplied by the caller, it seems duplicates should be permitted. Am I missing something? Is this an annotation processing constraint?

I've solved the problem by adding qualifying annotations to both the `MyClass` constructor declaration and the factory function.

**Comment by oldergod** [comment]
Order of parameters has no meaning for this lib.

See https://github.com/square/AssistedInject/blob/master/assisted-inject-processor/src/test/java/com/squareup/inject/assisted/processor/AssistedInjectProcessorTest.kt#L907-L953

**Comment by JakeWharton** [comment]
Can you explain the use case more? This was always intended as the behavior but we weren't validating it in 0.1.x. For now you can work around it by using qualifier annotations on the assisted types to disambiguate them.

**Comment by btilbrook-nextfaze** [comment]
If it's intended behavior, that's fine, and perhaps it's really a documentation issue. I am curious about why parameter ordering isn't considered/respected. When I write a factory by hand I am able to declare whatever provided args I need to create the object. I've raised this because as a consumer of the library, the restriction seems arbitrary.

**Comment by JakeWharton** [comment]
It's for symmetry with the behavior of the dependency injector and the arguments that it provides.

I'm not opposed to changing it but then we have to consider things like:
 * Do we force the ordering to always match?
 * How do we deal with parameter name mismatches between the factory and the constructor? (e.g., if you reversed the parameter naming of your original example what would you expect to happen?)
 * Do we prevent qualifiers from being used altogether?

**Comment by btilbrook-nextfaze** [comment]
> Do we force the ordering to always match?

I think it's quite reasonable. That was my expectation initially. What is actually _gained_ by supporting out of order args?

> if you reversed the parameter naming of your original example what would you expect to happen?

Just as if it were a factory by hand, I'd expect the programmer to take care in ensuring it matches the counterpart. It's not a new problem and I wouldn't expect the library to attempt to solve it (although it is a problem often solved using refactor tools, which aren't likely applicable here).

> Do we prevent qualifiers from being used altogether?

I assume you mean "Do we prevent qualifiers from being used altogether _for assisted args_?". If you agree the library doesn't need to solve the above arg order problem, then I'm not sure I see the need for qualifiers on assisted args any more. It seems this library aims to reduce boilerplate, and declaring the qualifying annotations to deal with this use case seems to be exactly that.

**Comment by JakeWharton** [comment]
I don't think we gain anything. The intent was just to start with the semantics of a regular injector where order isn't a thing.

For validation I'd be tempted to force the argument names to match only when you have more than one argument of the same type.

**Comment by JakeWharton** [comment]
Just released as 0.3.0. For implementation simplicity we actually force all argument names to match now. Please file a new issue if this is too much of a problem. I wanted to be too strict with the potential to relax should it prove too annoying.

**Comment by btilbrook-nextfaze** [comment]
Sounds good to me. Thanks for the change.
